### ISSUE TITLE
Move cloud snitches logic to scripts

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#12306]
There are 4 of them currently -- gce, ec2, ec2-multi-region and azure. And maybe will come more (or maybe not). Having them all sitting as built-in scylla code is not very flexible, hard to debug (#12185, #12160, #10250, #10494) and produces code duplication

Since it's not legal to change node's dc and rack on the fly (there's replace operation for that)  it's possible to replace all cloud snitches with gossiping-property-file one. In cloud environments the g.p.f. snitch config file should be prepared by scylla startup/install scripts and scylla will just use the provided info.

**Comment by xemul** [comment]
Cc @mykaul

**Comment by avelanarius** [comment]
> I feel that the drivers consuming the same format (gossiping-property-file) is probably going to help reduce maintenance as well.

Yes, it could be a solution, but it would have a strange user experience. First of all, on Scylla Cloud/serverless the users wouldn't have access to the gossiping-property-file (although maybe the Scylla Cloud could allow downloading it, but that's probably a wrong direction and unneccesary complexity). But even if they did, the user would have to start Scylla, SSH to a node, copy a file(s) and distribute it to all their API servers that actually use the driver.

**Comment by mykaul** [comment]
> > I feel that the drivers consuming the same format (gossiping-property-file) is probably going to help reduce maintenance as well.
> 
> Yes, it could be a solution, but it would have a strange user experience. First of all, on Scylla Cloud/serverless the users wouldn't have access to the gossiping-property-file (although maybe the Scylla Cloud could allow downloading it, but that's probably a wrong direction and unneccesary complexity). But even if they did, the user would have to start Scylla, SSH to a node, copy a file(s) and distribute it to all their API servers that actually use the driver.

My thought was that the driver will create the same format of the file (across drivers and what Scylla server uses)

**Comment by xemul** [comment]
Also note, that even now scylla configuration scripts _know_ the cloud provider it's running on and configure scylla.yaml (I think it's scylla.yaml) with respective snitch driver.

**Comment by mykaul** [comment]
While at it, make sure on EC2 we move to IMDSv2 - see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html for more information.

**Comment by mykaul** [comment]
@yaronkaikov , @avelanarius (for drivers) - let's see if we can fit this into Q3.

**Comment by syuu1228** [comment]
Sent draft version of the PRs to implement this, see the link above.

**Comment by tgrabiec** [comment]
> @tgrabiec - I assume in K8S you'll be able to provide this data via labels on the nodes?

Did you mean @tnozicka ?

**Comment by mykaul** [comment]
> > @tgrabiec - I assume in K8S you'll be able to provide this data via labels on the nodes?
> 
> Did you mean @tnozicka ?

Yes!

**Comment by syuu1228** [comment]
@xemul @mykaul Could you answer this? https://github.com/scylladb/scylladb/pull/13623#issuecomment-1518979277
I have assigned this issue but I don't really understand why we have to do this.

**Comment by mykaul** [comment]
I think the description catches it perfectly - what is coded in C++ and is hard to debug and fix (and certainly cannot be shared between the nodes and the various driver code!), can be easily scripted to create a static file for cloud topology - probably simple Python scripts can do the job, per snitch, or all of them in a single file, etc. The idea is to convert that C++ built-in code to external script that would run before launching Scylla (or a driver)

**Comment by tnozicka** [comment]
>> @tgrabiec - I assume in K8S you'll be able to provide this data via labels on the nodes?

> Did you mean @tnozicka ?

Sorry this got lost on me. We use `endpoint_snitch: "GossipingPropertyFileSnitch"` so I guess there wouldn't be a change needed on our side, right?

**Comment by mykaul** [comment]
> > > @tgrabiec - I assume in K8S you'll be able to provide this data via labels on the nodes?
> 
> > Did you mean @tnozicka ?
> 
> Sorry this got lost on me. We use `endpoint_snitch: "GossipingPropertyFileSnitch"` so I guess there wouldn't be a change needed on our side, right?

This is exactly what the script would do - create this file.

**Comment by mykaul** [comment]
> While at it, make sure on EC2 we move to IMDSv2 - see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html for more information.

This needs to get higher priority - https://aws.amazon.com/blogs/aws/amazon-ec2-instance-metadata-service-imdsv2-by-default/ - @yaronkaikov - do we have a specific item to support IMDSv2 in the current snitch?

**Comment by yaronkaikov** [comment]
> > While at it, make sure on EC2 we move to IMDSv2 - see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html for more information.
> 
> This needs to get higher priority - https://aws.amazon.com/blogs/aws/amazon-ec2-instance-metadata-service-imdsv2-by-default/ - @yaronkaikov - do we have a specific item to support IMDSv2 in the current snitch?

I am not familiar with that area. @syuu1228 Can you please answer?

**Comment by mykaul** [comment]
> [@xemul](https://github.com/xemul) - please create a PR leaving only yaml reading at startup. Please correlate with [@syuu1228](https://github.com/syuu1228) (I guess?) to make sure the yaml is actually created correctly just before starting Scylla.

There's no code yet (there is in Scylla cloud) for script based snitch setup.

**Comment by regevran** [comment]
We want to remove the wisdom from scylla as there is no functionality related issue - it is just updated in startup and stays forever. 
Removing it from Scylla's code will simplify the code and its maintenance.

**Comment by mykaul** [comment]
> We could move the logic to scripts, or move it into main() to reduce the dependency on external stuff.
> 
> Although, the scylla binary wouldn't have permissions into /etc.
> 
> btw, are the various snitch output disjoint? If not, we can't create a cross-cloud-provider cluster. It would have been better for each snitch to prefix its output with the cloud provider name.

The idea of moving them outside is also to allow the client to use them - the client should know in which AZ/region they are as well, to connect to the 'local' AZ.

**Comment by mykaul** [comment]
> [@tzach](https://github.com/tzach) [@avikivity](https://github.com/avikivity) as per our summit talk to increase priority of this
> I guess we could reuse this bug to prioritize getting rid of snitches.

I'm not aware of this conversation and I'm not sure we have the resources to do it. It certainly wasn't planned. I'm in favor of the effort of course.
